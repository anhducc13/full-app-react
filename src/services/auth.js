import axios from 'axios';
const URL = process.env.REACT_APP_API_URL

const register = params => {
    return axios.post(`${URL}/auth/register`, params)
}

const verify = (accessToken) => {
    return axios.get(`${URL}/auth/verify/${accessToken}`)
}

const login = (params) => {
    return axios.post(`${URL}/auth/login`, params)
}

const logout = (accessToken) => {
    const reqConfig = {
        "headers": {
            "Authorization": `Bearer ${accessToken}`,
        }

    }
    return axios.get(`${URL}/auth/logout`, reqConfig)
}

const updatePassword = (params, accessToken) => {
    const reqConfig = {
        "headers": {
            "Authorization": `Bearer ${accessToken}`,
        }
    }
    return axios.post(`${URL}/auth/updatePassword`, params, reqConfig)
}

const forgotPassword = params => {
    return axios.post(`${URL}/auth/forgotPassword`, params)
}

export default {
    login,
    logout,
    register,
    verify,
    updatePassword,
    forgotPassword
}