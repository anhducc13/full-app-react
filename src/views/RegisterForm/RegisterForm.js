import React from 'react';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import TextField from '@material-ui/core/TextField';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink} from 'react-router-dom'

const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(3),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));

export const RegisterForm = (props) => {
  const classes = useStyles();

  const {
    emailField,
    usernameField,
    passwordField,
    confirmPasswordField,
  } = props.textField;

  const {
    onChangeEmail,
    onChangeUsername,
    onChangePassword,
    onChangeConfirmPassword,
  } = props.handleChange

  const submitForm = props.handleSubmit

  return (
    <Container component="main" maxWidth="sm">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Register
        </Typography>
        <form className={classes.form} onSubmit={submitForm} noValidate>
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="email"
            label="Email"
            name="email"
            autoFocus
            value={emailField.value}
            error={emailField.error !== ''}
            onChange={onChangeEmail}
            helperText={emailField.valid ? "" : emailField.error}
          />
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="username"
            label="Username"
            name="username"
            value={usernameField.value}
            error={usernameField.error !== ''}
            onChange={onChangeUsername}
            helperText={usernameField.valid ? "" : usernameField.error}
          />
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            name="password"
            label="Password"
            type="password"
            id="password"
            value={passwordField.value}
            error={passwordField.error !== ''}
            onChange={onChangePassword}
            helperText={passwordField.valid ? "" : passwordField.error}
          />
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            name="confirm-password"
            label="Confirm Password"
            type="confirm-password"
            id="confirm-password"
            value={confirmPasswordField.value}
            error={confirmPasswordField.error !== ''}
            onChange={onChangeConfirmPassword}
            helperText={confirmPasswordField.valid ? "" : confirmPasswordField.error}
          />
          <Button
            size="large"
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
            disabled={
              !emailField.valid || !usernameField.valid
              || !passwordField.valid || !confirmPasswordField.valid
            }
          >
            Register
          </Button>
          <Grid container>
            <Link to="/login" variant="body2" component={RouteLink}>
              Have an account? Login
            </Link>
          </Grid>
        </form>
      </div>
    </Container>
  );
}