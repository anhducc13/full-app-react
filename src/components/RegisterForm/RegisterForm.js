import React, { useState } from 'react';
import { setGlobal } from 'reactn';
import Avatar from '@material-ui/core/Avatar';
import CssBaseline from '@material-ui/core/CssBaseline';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink } from 'react-router-dom';
import { Button as ButtonCustom } from '../shared/Button';
import { InputText } from '../shared/InputText';
import Swal from 'sweetalert2';
import { authService } from 'services';
import { validateUsername, validatePassword, validateEmail } from 'helpers/validators'


const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(3),
  },
}));

export const RegisterForm = (props) => {
  const classes = useStyles();

  const initTextField = {
    value: '',
    error: '',
    valid: false,
  };
  const [usernameField, setUsernameField] = useState(initTextField);
  const [emailField, setEmailField] = useState(initTextField);
  const [passwordField, setPasswordField] = useState(initTextField);
  const [confirmPasswordField, setConfirmPasswordField] = useState(initTextField);

  const onChangeEmail = (e) => {
    let inputValidate = Object.assign(emailField);
    inputValidate.value = e.target.value;
    const outputValidate = validateEmail(inputValidate);
    setEmailField(outputValidate);
  }

  const onChangeUsername = (e) => {
    let inputValidate = Object.assign(usernameField);
    inputValidate.value = e.target.value;
    const outputValidate = validateUsername(inputValidate);
    setUsernameField(outputValidate);
  }

  const onChangePassword = (e) => {
    let inputValidate = Object.assign(passwordField);
    inputValidate.value = e.target.value;
    const outputValidate = validatePassword(inputValidate);
    setPasswordField(outputValidate);
  }

  const onChangeConfirmPassword = (e) => {
    const { password } = passwordField;
    let inputValidate = Object.assign(confirmPasswordField);
    inputValidate.value = e.target.value;

    if (inputValidate.value !== password.value) {
      inputValidate.error = 'Mật khẩu không khớp!';
      inputValidate.valid = false;
    } else {
      inputValidate.error = '';
      inputValidate.valid = true;
    }
    setConfirmPasswordField(inputValidate);
  }

  const onSubmitForm = (e) => {
    e.preventDefault()
    const payload = {
      "username": usernameField.value,
      "email": emailField.value,
      "password": passwordField.value,
    }
    setGlobal({
      loading: true,
    })
    authService.register(payload)
      .then(() => {
        setGlobal({
          loading: false,
        })
        Swal.fire(
          'Thành công!',
          'Vui lòng kiểm tra email xác thực tài khoản để có thể đăng nhập vào trang web',
          'success'
        )
        props.history.push('/')
      })
      .catch(err => {
        setGlobal({
          loading: false,
        })
        if (err.response && err.response.status === 400) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Dữ liệu đầu vào có vấn đề',
            'error'
          )
        }
        else if (err.response && err.response.status === 409) {
          Swal.fire(
            'Có lỗi xảy ra!',
            'Username hoặc email đã được đăng ký trước đó',
            'error'
          )
        } else {
          props.history.push('/error')
        }
      })
  }

  return (
    <Container component="main" maxWidth="sm">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Register
        </Typography>
        <form className={classes.form} noValidate onSubmit={onSubmitForm}>
          <InputText
            textField={emailField}
            name="email"
            label="Email"
            type="email"
            onChange={onChangeEmail}
          />
          <InputText
            textField={usernameField}
            name="username"
            label="Username"
            type="text"
            onChange={onChangeUsername}
          />
          <InputText
            textField={passwordField}
            name="password"
            label="Password"
            type="password"
            onChange={onChangePassword}
          />
          <InputText
            textField={confirmPasswordField}
            name="confirm-password"
            label="Confirm Password"
            type="password"
            onChange={onChangeConfirmPassword}
          />
          <ButtonCustom
            displayText="Register"
            disabled={
              !usernameField.valid || !passwordField.valid ||
              !emailField.valid || !confirmPasswordField.valid
            }
          />

          <Grid container>
            <Link to="/login" variant="body2" component={RouteLink}>
              Have an account? Login
            </Link>
          </Grid>
        </form>
      </div>
    </Container>
  );
}