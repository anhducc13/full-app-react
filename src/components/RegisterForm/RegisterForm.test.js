import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { RegisterForm } from './RegisterForm';
import { authService } from '../../services';

describe('Register Form Render', () => {
  it('should render correctly', () => {
    const component = shallow(<RegisterForm />);
    expect(component).toMatchSnapshot();
  });
});

describe('Register Form validate email', () => {
  it('validate email test case 1', () => {
    const component = mount(
      <MemoryRouter>
        <RegisterForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: ""
      }
    }
    component.find('input[name="email"]').simulate('change', targetInput);
    expect(component.find('p#email-helper-text').text())
      .toBe('Email không hợp lệ');
    component.unmount()
  })

  it('validate email test case 2', () => {
    const component = mount(
      <MemoryRouter>
        <RegisterForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: " "
      }
    }
    component.find('input[name="email"]').simulate('change', targetInput);
    expect(component.find('p#email-helper-text').text())
      .toBe('Email không hợp lệ');
    component.unmount()
  })

  it('validate email test case 3', () => {
    const component = mount(
      <MemoryRouter>
        <RegisterForm />
      </MemoryRouter>
    );
    const targetInput = {
      target: {
        value: "trantienduc10@gmail.com"
      }
    }
    component.find('input[name="email"]').simulate('change', targetInput);
    expect(component.find('p#email-helper-text').length).toBe(0)
    component.unmount()
  })
});

describe('Register form submit', () => {
  it('Button submit disable with invalid data', () => {
    const component = mount(
      <MemoryRouter>
        <RegisterForm />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: '' } });
    expect(component.find('button[type="submit"]').prop('disabled')).toBeTruthy();
    component.unmount();
  })

  it('Button submit enable with expect data', () => {
    const component = mount(
      <MemoryRouter>
        <RegisterForm />
      </MemoryRouter>
    );
    component.find('input[name="username"]').simulate('change', { target: { value: 'anhducc13' } });
    component.find('input[name="email"]').simulate('change', { target: { value: 'trantienduc10@gmail.com' } });
    component.find('input[name="password"]').simulate('change', { target: { value: 'Anhducc13' } });
    component.find('input[name="confirm-password"]').simulate('change', { target: { value: 'Anhducc13' } });
    expect(component.find('button[type="submit"]').prop('disabled')).toBeFalsy()
    component.unmount();
  })

  it('Submit form, call api with valid input', () => {
    const component = mount(
      <MemoryRouter>
        <RegisterForm />
      </MemoryRouter>
    );
    const value = {
      'username': 'anhducc13',
      'password': 'Anhducc13',
      'email': 'trantienduc10@gmail.com'
    }
    component.find('input[name="username"]').simulate('change', { target: { value: value.username } });
    component.find('input[name="email"]').simulate('change', { target: { value: value.email } });
    component.find('input[name="password"]').simulate('change', { target: { value: value.password } });
    const registerFn = jest.spyOn(authService, 'register').mockImplementation(() => Promise.resolve({ status: 200 }));
    component.find('form').simulate('submit');
    expect(registerFn).toHaveBeenCalled();
    expect(registerFn).toHaveBeenCalledWith(value);
    component.unmount();
  })
});
