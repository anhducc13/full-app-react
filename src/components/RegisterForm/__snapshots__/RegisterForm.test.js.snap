// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Form Render should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  component="main"
  maxWidth="sm"
>
  <CssBaseline />
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(Avatar))
      className="makeStyles-avatar-2"
    >
      <LockOutlinedIcon />
    </WithStyles(ForwardRef(Avatar))>
    <WithStyles(ForwardRef(Typography))
      component="h1"
      variant="h4"
    >
      Đăng ký
    </WithStyles(ForwardRef(Typography))>
    <form
      className="makeStyles-form-3"
      noValidate={true}
      onSubmit={[Function]}
    >
      <InputText
        label="Email"
        name="email"
        onChange={[Function]}
        textField={
          Object {
            "error": "",
            "valid": false,
            "value": "",
          }
        }
        type="email"
      />
      <InputText
        label="Tên đăng nhập"
        name="username"
        onChange={[Function]}
        textField={
          Object {
            "error": "",
            "valid": false,
            "value": "",
          }
        }
        type="text"
      />
      <InputText
        label="Mật khẩu"
        name="password"
        onChange={[Function]}
        textField={
          Object {
            "error": "",
            "valid": false,
            "value": "",
          }
        }
        type="password"
      />
      <InputText
        label="Nhập lại mật khẩu"
        name="confirm-password"
        onChange={[Function]}
        textField={
          Object {
            "error": "",
            "valid": false,
            "value": "",
          }
        }
        type="password"
      />
      <Button
        disabled={true}
        displayText="Đăng ký"
      />
      <WithStyles(ForwardRef(Grid))
        container={true}
      >
        <WithStyles(ForwardRef(Link))
          component={[Function]}
          to="/dang-nhap"
          variant="body2"
        >
          Đã có tài khoản? Đăng nhập
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Grid))>
    </form>
  </div>
</WithStyles(ForwardRef(Container))>
`;
